cmd/ipfs/daemon.go

```diff
@@ -461,3 +482,80 @@ func merge(cs ...<-chan error) <-chan error {
 	}()
 	return out
 }
+
+func runGC(req cmds.Request, repo repo.Repo, node *core.IpfsNode) error {
+	disableGC, _, err := req.Option(disableGCKwd).Bool()
+	if err != nil {
+		return err
+	}
+	if disableGC {
+		return nil
+	}
+
+	cfg, err := req.InvocContext().GetConfig()
+	if err != nil {
+		return fmt.Errorf("runGC: GetConfig() failed: %s", err)
+	}
+	ctx := req.Context()
+
+	// check if cfg has these fields initialized
+	// TODO: there should be a general check for all of the cfg fields
+	// maybe distinguish between user config file and default struct?
+	if cfg.Datastore.StorageMax == 0 {
+		repo.SetConfigKey("Datastore.StorageMax", 10*1000)
+		cfg.Datastore.StorageMax = 10 * 1000
+	}
+	if cfg.Datastore.StorageGCWatermark == 0 {
+		repo.SetConfigKey("Datastore.StorageGCWatermark", 90)
+		cfg.Datastore.StorageGCWatermark = 90
+	}
+	if cfg.Datastore.GCTimeoutMinute == 0 {
+		repo.SetConfigKey("Datastore.GCTimeoutMinute", 60)
+		cfg.Datastore.GCTimeoutMinute = 60
+	}
+
+	// calculate "swap" size, the slack space between StorageMax and StorageGCWatermark
+	// used to limit GC duration
+	swapGB := cfg.Datastore.StorageMax / 1000 * (100 - cfg.Datastore.StorageGCWatermark) / 100
+	if swapGB < 1 {
+		swapGB = 1
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(time.Duration(cfg.Datastore.GCTimeoutMinute) * time.Minute):
+			// Check fs repo disk usage
+			diskUsage, err := repo.GetDiskUsage()
+			if err != nil {
+				return err
+			}
+			diskUsageMB := diskUsage / (1000 * 1000)
+			if diskUsageMB < 1 {
+				// round up to 1 MB
+				diskUsageMB = 1
+			}
+
+			if diskUsageMB > cfg.Datastore.StorageGCWatermark*cfg.Datastore.StorageMax/100 {
+				// Do GC here
+				fmt.Println("Starting repo GC...")
```

Comment:
there's a few `fmt.Println`, sorry, those are very few and mostly just the addresses users see when starting up the daemon. 

we've considered maybe printing out some of the events, kind of like an http server, but it will take some work to figure out what to show and what not to.